home/utils.py

```diff
@@ -217,86 +248,338 @@ def add_viber(row, page=None):
+    # Padding
+    sheet.insert_cols(1)
+
+    # Colours
+    blue = Color(rgb="0099CCFF")
+
+    # Boarders
+    right_border = Border(right=Side(border_style="thin", color="FF000000"))
+
+    # Fills
+    blue_fill = PatternFill(patternType="solid", fgColor=blue)
+
+    # Named Styles
+    header_style = NamedStyle(name="header_style")
+    menu_style = NamedStyle(name="menu_style")
+
+    # Set attributes to styles
+    header_style.font = Font(bold=True, size=12)
+    menu_style.fill = blue_fill
+    menu_style.font = Font(bold=True, size=10)
+
+    # Add named styles to wb
+    wb.add_named_style(header_style)
+    wb.add_named_style(menu_style)
+
+    # Set menu style for any "Menu" row
+    for row in sheet.iter_rows():
+        if isinstance(row[1].value, str) and "Menu" in row[1].value:
+            for cell in row:
+                cell.style = menu_style
+
+    # Set header style for row 1 and 2
+    for row in sheet["1:2"]:
+        for cell in row:
+            cell.style = header_style
+
+    # Set dividing border for side panel
+    for cell in sheet["G:G"]:
+        cell.border = right_border
+    return wb, sheet
+
+
+def set_level_numbers(rows: List[list]) -> List[list]:
+    """Sets the level number in the side panel to indicate depth of the page in a visual way"""
+    menu = 0
+    sub_1 = 0
+    sub_2 = 0
+    sub_3 = 0
+    sub_4 = 0
+    for row in rows:
+        if row[0] == "x":
+            if row[5] == 1:
+                menu += 1
+                row[0] = f"Menu {menu}"
+                sub_1 = 0
+                sub_2 = 0
+                sub_3 = 0
+                sub_4 = 0
+            else:
+                row[0] = ""
+        elif row[1] == "x":
+            if row[5] == 1:
+                sub_1 += 1
+                row[1] = f"Sub {menu}.{sub_1}"
+                sub_2 = 0
+                sub_3 = 0
+                sub_4 = 0
+            else:
+                row[1] = ""
+        elif row[2] == "x":
+            if row[5] == 1:
+                sub_2 += 1
+                row[2] = f"Sub {menu}.{sub_1}.{sub_2}"
+                sub_3 = 0
+                sub_4 = 0
+            else:
+                row[2] = ""
+        elif row[3] == "x":
+            if row[5] == 1:
+                sub_3 += 1
+                row[3] = f"Sub {menu}.{sub_1}.{sub_2}.{sub_3}"
+                sub_4 = 0
+            else:
+                row[3] = ""
+        elif row[4] == "x":
+            if row[5] == 1:
+                sub_4 += 1
+                row[4] = f"Sub {menu}.{sub_1}.{sub_2}.{sub_3}.{sub_4}"
+            else:
+                row[4] = ""
+    return rows
+
+
+def get_rows(page: ContentPage) -> List[list]:
+    """Sets up row for each page including the side panel.
+    Each page is returned as a list of rows"""
+    actual_depth = (
+        page.depth - 3
+    )  # main menu is depth 3 (is this always true? root -> home -> main menu)
+    padding = [""] * 6
+    padding[actual_depth] = "x"
```

Comment:
I'm wondering if there's a better way of doing this, rather than setting `x`'s and then overwriting them with the correct value later, couldn't we set them to the correct value here? (yet again, that looks like a big change, so probably not for this PR).

Reply:
the reason I decided to go with the x's and set them later was because SXD wants these long strings with lots of points and numbers that depend on the parent as well as other children and at the moment, the row has no awareness of what came before, so I thought it might be easier (albeit way uglier) to iterate over the "sheet" afterwards and set the numbering strings. I am really not sure though, I did consider adding a value that gets passed into the `add_children` function but I am really not sure, I think it might be an option to look into again when/if the dataclass gets implemented?

Reply:
Yeah, definitely happy for it not to happen now. Maybe the dataclass can include it's "position" (parent, depth, etc), and then when iterating through the rows we can easily know what to put here.